mint-client/src/clients/user.rs

```diff
@@ -323,6 +335,65 @@ impl UserClient {
             .await
             .map_err(|_| ClientError::WaitContractTimeout)?
     }
+
+    pub async fn create_invoice_and_offer<R: RngCore + CryptoRng>(
+        &self,
+        amount: Amount,
+        mut rng: R,
+    ) -> Result<(KeyPair, Invoice), ClientError> {
+        let (payment_keypair, payment_public_key) =
+            self.context.secp.generate_schnorrsig_keypair(&mut rng);
+        let raw_payment_secret = payment_public_key.serialize();
+        let payment_hash = bitcoin::secp256k1::hashes::sha256::Hash::hash(&raw_payment_secret);
+        let payment_secret = PaymentSecret(raw_payment_secret);
+
+        // Temporary lightning node pubkey
+        let (node_secret_key, node_public_key) = self.context.secp.generate_keypair(&mut rng);
+
+        let invoice = InvoiceBuilder::new(network_to_currency(self.context.config.wallet.network))
```

Comment:
i would expect the lightning gateway node to be responsible for generating the actual invoice, but it seems like the user here (who is not running a real lightning node) is creating the invoice?

Reply:
Indeed! If the LN GW was generating the invoice it would need to be trusted because it would also generate the preimage. Instead we let the user generate the preimage and the LN GW is only the last hop (and is advertised as a route hint in the invoice) and not the recipient. Instead the federation is the recipient on behalf of the user (it's trusted anyway) and they can claim their funds when they are online.

Reply:
Gotcha thanks. I was confused because I didn't see the route hint in this PR, but now knowing the route hint will be there it makes sense :)

Reply:
The mechanism I was thinking of is the user still generates the pre-image, and gives the payment hash to the Gateway node, but then the Gateway node generates the actual invoice using that payment hash (without knowing the pre-image)

Then once the payment htlc comes in, the Gateway Node can purchase the preimage w/ ecash tokens to settle the payment.

It's a similar mechanism to what you described, trying to think if there would be any benefits or things I'm missing. One benefit would be smaller invoices (no need for the route hint)

Reply:
I see the route hint in this pr: https://github.com/fedimint/minimint/pull/101/files

Reply:
I was just considering the integration tests environment for this PR. Since there's no routing in this context, I didn't think it would make sense to include route hints. I have another branch that builds on this which has a working plugin https://github.com/justinmoon/minimint/pull/4. Once this hackathon is done I'll finish off this PR. Sorry for delay ...,

Reply:
I'm marking this resolved. Will add route hint in the next PR which adds core-lightning plugin and integration tests showing that the route-hint actually works ...